sound/soc/sof/ipc4-telemetry.c

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2018-2023 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry.h"
+#include "ipc4-priv.h"
+
+static void *sof_ipc4_query_exception_address(struct snd_sof_dev *sdev, void *cached_addr)
+{
+	u32 telemetry_slot_offset;
+	void __iomem *io_addr;
+	void *start_addr;
+	u32 offset;
+
+	telemetry_slot_offset =
+		sof_ipc4_find_debug_slot_offset_by_type(sdev, cached_addr,
+							SOF_IPC4_DEBUG_SLOT_TELEMETRY);
+	if (!telemetry_slot_offset)
+		return NULL;
+
+	/* skip the first separator code */
+	offset = telemetry_slot_offset + sizeof(u32);
+
+	if (!pm_runtime_active(sdev->dev)) {
```

Comment:
I really don't like all these tests for pm_runtime_active. This is super racy and error-prone

Can we think of something more reliable?

Reply:
removed